### PR TITLE
Updates for Ubuntu-Pro with magic-attach

### PR DESCRIPTION
**Pull request by ogayot**
Apologies for the size of the diff :/

This adds the support for magic-attach / contract-selection in Subiquity.

Unfortunately, testing this PR on a current ubuntu server ISO will fail because we are relying on an unreleased feature of ubuntu-advantage-tools. Also the ubuntu.com/pro/attach website is not yet available for the user to do the contract selection.

The dry-run implementation exists but is limited. 
Upon reaching the Ubuntu Pro screen, the user is presented the user-code `AAAAAA`.
The user-code expires after 10 minutes (affected by `SUBIQUITY_REPLAY_TIMESCALE`).
When the user-code expires, a new contract selection is initiated and the next user-code (`AAAAAB`) will be presented instead,
The user has no way to select a contract in dry-run mode so the user-code will keep expiring until the user inputs a contract token instead.

More integration testing can be done by following the steps below:

* build the development version of ubuntu-advantage-tools
  * checkout https://github.com/canonical/ubuntu-advantage-client.git
  * `sbuild -d kinetic`
* deploy a LXD container (ubuntu 22.04 for instance)
* install the .deb built previously on the container
* apply the following diff to subiquity (after replacing "pleasant-griffon" with the name of the LXD container):
```diff
diff --git a/subiquity/server/controllers/ubuntu_pro.py b/subiquity/server/controllers/ubuntu_pro.py
index 831ae635..5e2c0f39 100644
--- a/subiquity/server/controllers/ubuntu_pro.py
+++ b/subiquity/server/controllers/ubuntu_pro.py
@@ -89,15 +89,11 @@ class UbuntuProController(SubiquityController):
     def __init__(self, app) -> None:
         """ Initializer for server-side Ubuntu Pro controller. """
         strategy: UAInterfaceStrategy
-        if app.opts.dry_run:
-            strategy = MockedUAInterfaceStrategy(scale_factor=app.scale_factor)
-        else:
-            # Make sure we execute `$PYTHON "$SNAP/usr/bin/ubuntu-advantage"`.
-            executable = (
-                os.environ["PYTHON"],
-                os.path.join(os.environ["SNAP"], "usr/bin/ubuntu-advantage"),
-            )
-            strategy = UAClientUAInterfaceStrategy(executable=executable)
+        # Make sure we execute `$PYTHON "$SNAP/usr/bin/ubuntu-advantage"`.
+        executable = (
+            "lxc", "exec", "pleasant-griffon", "--", "/usr/bin/ubuntu-advantage",
+        )
+        strategy = UAClientUAInterfaceStrategy(executable=executable)
         self.ua_interface = UAInterface(strategy)
         self.cs: Optional[ContractSelection] = None
         self.magic_token = Optional[str]
```
* the contracts CLI tool can be used to simulate what the user could do on ubuntu.com/pro/attach

### CODE REVIEW

File: subiquity/server/contract_selection.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+        return cls(
+                client=client,
+                magic_token=answer["data"]["attributes"]["token"],
+                validity_seconds=answer["data"]["attributes"]["expires_in"],
+                user_code=answer["data"]["attributes"]["user_code"])
+
+    async def _run_polling(self) -> str:
+        """ Runs the polling and eventually return a contract token. """
+        # Wait an initlal 30 seconds before sending the first request, then
```
Comment (dbungert) [inline subiquity/server/contract_selection.py:77]: minor typo

File: subiquity/ui/views/ubuntu_pro.py [hunk 2]
```diff
@@ -290,7 +391,12 @@ def upgrade_yes_no_screen(self) -> Widget:
         +---------------------------------------------------------+
         """
 
-        excerpt = _("Upgrade this machine to Ubuntu Pro or skip this step.")
+        excerpt = _("Upgrade this machine to Ubuntu Pro for security updates"
+                    " on a much wider range of packages, until 2032. Assists"
```
Comment (dbungert) [inline subiquity/ui/views/ubuntu_pro.py:395]: currentyear + 10?

File: subiquity/ui/views/ubuntu_pro.py [hunk 3]
```diff
@@ -598,16 +759,17 @@ def __init__(self, parent: UbuntuProView) -> None:
...
 
         services = [
-            _("Security patching for over 30.000 packages, with a focus on"
-              " High and Critical CVEs (extended from 2.500)"),
-            _("10 years of security Maintenance (extended from 5 years)"),
-            _("Kernel Livepatch service for increased uptime and security"),
-            _("Ubuntu Security Guide for hardening profiles, including CIS"
-              " and DISA-STIG"),
+            _("Security patch coverage for CVSS critical, high and selected"
+              ' medium vulnerabilities in all 23,000 packages in "universe"'
```
Comment (dbungert) [inline subiquity/ui/views/ubuntu_pro.py:768]: Would be nice to pull these numbers out of the translated string, they are likely to change.

File: subiquity/tests/api/test_api.py [hunk 4]
```diff
@@ -1320,3 +1320,30 @@ async def test_create_primaries(self):
             }
             with self.assertRaises(ClientResponseError):
                 await inst.post('/storage/v2/add_partition', data)
+
+
+class TestUbuntuProContractSelection(TestAPI):
+    async def test_upcs_flow(self):
```
Comment (dbungert) [inline subiquity/tests/api/test_api.py:1326]: `@timeout` please

File: subiquity/server/contract_selection.py (hunk 1 above)
Comment (ogayot) [reply subiquity/server/contract_selection.py:77]: Nice catch, I couldn't see it ever after you said it :+1:

File: subiquity/ui/views/ubuntu_pro.py (hunk 3 above)
Comment (ogayot) [reply subiquity/ui/views/ubuntu_pro.py]: Yep, makes sense. Ideally, those numbers should be dependent on the version of Ubuntu being installed but I don't see a simple way to do that.

File: subiquity/ui/views/ubuntu_pro.py (hunk 2 above)
Comment (ogayot) [reply subiquity/ui/views/ubuntu_pro.py:395]: I feel like that would give a wrong information to the user, wouldn't it?

If somebody attempts an install of Jammy in say 2025, he would get "until 2035". Sadly, this is probably another instance of an information that should be dependent on the release being installed.

Comment (dbungert) [reply subiquity/ui/views/ubuntu_pro.py:395]: /usr/share/distro-info/ubuntu.csv from package distro-info-data ?

Comment (dbungert) [reply subiquity/ui/views/ubuntu_pro.py:395]: To not complicate this PR further, let's pull the number out of the string and deal with getting better numbers in a future PR.

**Review by dbungert — APPROVED** [review]

**Pull request merged**